python/polars_ds/num.py

```diff
@@ -1066,3 +1066,43 @@ def _haversine(
             is_elementwise=True,
             cast_to_supertypes=True,
         )
+
+    def knn_entropy(self, *others: pl.Expr, k=2):
+        """
+        Multivariate estimation of entropy through a k-nearest neighbor query
+
+        Reference
+        ---------
+        https://arxiv.org/pdf/1506.06501v1.pdf
+        """
+        from scipy.special import digamma
```

Comment:
I think I might have a rust digamma function available somewhere already in the Rust codebase. If not, we can use the digamma implemented here: https://github.com/statrs-dev/statrs/blob/5411ba74b427b992dd1410d699052a0b41dc2b5c/src/function/gamma.rs#L353. 

One of my goal is to reduce dependency on other packages as much as possible. (But I personally admit the crime that I have way tooooo many packages installed in my env...)

Reply:
+1 to reuse internal functions.
In this case, does that mean copying/pasting code from your link into `gamma.rs` in this repo ?

Reply:
We can. The statsrs project has MIT license, and I am acknowledging them in this project. 

The question is whether we want to expose this digamma function in the Rust layer or the Python layer. If it is only in the Rust layer, we might need to bring this KNN into Rust and the function should return something like self._expr.register_plugin.....

It is up to you.

Reply:
found this https://gist.github.com/timvieira/656d9c74ac5f82f596921aa20ecb6cc8
seems really helpful for an implementation owned by this repo.
The cython version mentioned could be reimplemented in Rust

@abstractqqq for this PR I was thinking about staying in pure Python space. I am a beginner in rust, so I would go one step at a time and keep this for latter

Reply:
So now for this PR we have two options: (1) Use the python code, name it "_digamma" or something and use it in your expression. Put a note linking to this issue (2) Wait till I write a Rust function to wrap it, and maybe u can call it via pld._diagamma or something.

I think I can get (2) done in maybe 10 minutes. But then again the performance gain is minimal and insignificant. Here is why I am a little concerned about (2). Doing (2) will mean that as a package, we will start providing functions outside of Polars dataframes. This is eventually planned, but I would like to do things in a more planned and organized manner... It's funny that the time I spent writing this comment is likely about the same as writing the Rust impl for digamma.. But I guess it shows my commitment to taking small steps and not adding sporadic functions/features to the package at the moment lol..

Reply:
@abstractqqq I'll follow your lead on this as I totally understand your concern.
I'll take my chance on (2). I can't guarantee to deliver under 10 minutes though ! 
Thanks for the time you spend on reviewing this

Reply:
It will be great!  Learning how to build Rust and Python interop is a lot of fun :) and can really help if you need high performance...

That said, if you build the digamma function in Rust, let's put it in utils (or gamma.rs in stats_utils), and maybe expose it in Python like from _polars_ds.stats_utils import _digamma or something like this.. You can find details on how to config the Python-side module names here. [Pyo3](https://pyo3.rs/v0.20.3/)